wagtail/blocks/list_block.py

```diff
@@ -224,6 +225,35 @@ def clean(self, value):
 
         return ListValue(self, bound_blocks=result)
 
+    def normalize(self, value):
+        if isinstance(value, ListValue):
+            return value
+
+        # Squint a little and see if it looks like a list
+        try:
+            items, _items = itertools.tee(value, 2)
+        except TypeError:
+            raise BlockNormalizationError(
+                f"Cannot handle {value!r} (type {type(value)!r}) as a value of a ListBlock"
+            )
+        try:
+            head = next(_items)
+        except StopIteration:
+            return self.empty_value()
+
+        if self._item_is_in_block_format(head):
+            # It looks like the json-ish representation
+            return ListValue(
+                self, values=[self.child_block.normalize(x["value"]) for x in items]
```

Comment:
This will discard any existing block IDs. However, rather than try to replicate the `to_python` logic to address this, I think we can reasonably leave this case out. We don't want `normalize` to be _too_ powerful, lest developers end up just throwing all data representations at it instead of using the proper conversion method - and I don't think there's any practical reason for people to be using the JSON-ish representation except when explicitly getting data in and out of the database.

We probably do still want to support `page.body = "some json string"`, but I think that's best handled at the top-level StreamField.to_python after all, and should be moved back there.

Reply:
Digging a bit deeper into this: initially it appeared that `StreamField.to_python` was called when unpacking a value from the database, and [`Creator.__set__`](https://github.com/wagtail/wagtail/blob/e653a4afbb7f61e0e27e74e60a96981ca1d27f60/wagtail/fields.py#L81) was the path taken when assigning to the field from Python code - which would have allowed us to keep these conversions entirely separate. However, Django also [calls `Field.to_python` during `Field.clean`](https://github.com/django/django/blob/5.0.4/django/db/models/fields/__init__.py#L829), which means that `Field.to_python` has to accept Python-ish values as well as JSON-ish ones.

In the old version of this code, `Field.to_python` performed value sniffing to distinguish between these cases, so `StreamBlock.to_python` only has to handle JSON-ish values:

https://github.com/wagtail/wagtail/blob/e653a4afbb7f61e0e27e74e60a96981ca1d27f60/wagtail/fields.py#L167-L172

In the new code, this logic is moved to `StreamBlock.normalize`, which doesn't seem quite right as it means that `StreamBlock.to_python` is taken out of the picture. However, if we're aiming to eliminate all StreamBlock-specific logic from `StreamField.to_python`, it does make sense to offload this to the block class - probably to `StreamBlock.to_python`.

I'm heading towards the conclusion that Django is forcing us to make `Block.to_python` into a "handle any type we throw at it" method, and thus there's no benefit to having a distinct `normalize` method after all (since anything that might benefit from it could just as well call `to_python` instead). However, I'll need to think some more about whether that's an API-breaking change.

Reply:
OK, that won't work without breaking changes. Here's what hopefully will, though...

Definitions:

* JSONish value: a block value as it appears in the database JSON representation, e.g. `[{"type": "...", "id": "...", "value": ...}]` for a StreamBlock
* Pythonic value: any value for the block that might be passed from Python code, such as the tuple representation or StreamValue for StreamBlock, or a string or RichText for RichTextBlock
* Normalized pythonic value: the value in its "correct" type as seen when we read the value back from the StreamField, such as a StreamValue or RichText

In the current codebase, `Block.to_python` must accept a JSONish value and return a normalized pythonic value. (If it needs to recurse over children, it calls `to_python` on them.)

As of this PR, `Block.normalize` must accept a Pythonic value and return a normalized pythonic value. (If it needs to recurse over children, it calls `normalize` on them.)

Add a new requirement: for a block to be valid at the top level of a StreamField (currently only the case for StreamBlock), its `to_python` method must recognise when it has been passed a Pythonic value, and handle it equivalently to `normalize`. (Since `to_python`'s functionality is effectively a superset of `normalize`, it's likely that such a block would just implement a `to_python` method that accepts both Pythonic and JSONish values, and alias `normalize` to it.)

Now, `StreamField.to_python` can and `Creator.__set__` can simply delegate to the top-level block's `.to_python` method.

Reply:
If `normalize` is implemented as an alias of `to_python`, is there any value to keeping `normalize` at all? It looks like most of the machinery for recursively calling `to_python` on subvalues is in place, the exception being `StreamBlock` which defers to `StreamValue`, which may or may not call `to_python` on subvalues, depending on the `is_lazy` argument.

Reply:
I think we do need to keep `normalize`, to account for existing block classes in non-Wagtail code that have a `to_python` method that only accepts JSONish values. (These classes won't implement `normalize`, so they'll just use the base `normalize` that returns the value unchanged.) If we don't have a separate `normalize` endpoint, then a container block handling a pythonic value and recursing over children will end up calling the child's `to_python` with a pythonic value, which isn't correct.

(A heads-up that I'm also working on this bit right now, so you might want to hold off to save duplicating work :-) )

Reply:
Done in https://github.com/gasman/wagtail/commit/e1ec5a3caf964035d137a3bd3be3e9811cd92129